flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/offset/BinlogOffset.java

```diff
@@ -0,0 +1,266 @@
+    public static final String BINLOG_POSITION_OFFSET_KEY = "pos";
+    public static final String EVENTS_TO_SKIP_OFFSET_KEY = "event";
+    public static final String ROWS_TO_SKIP_OFFSET_KEY = "row";
+    public static final String GTID_SET_KEY = "gtids";
+    public static final String TIMESTAMP_KEY = "ts_sec";
+    public static final String SERVER_ID_KEY = "server_id";
+
+    public static final BinlogOffset INITIAL_OFFSET = new BinlogOffset("", 0);
+    public static final BinlogOffset NO_STOPPING_OFFSET = new BinlogOffset("", Long.MIN_VALUE);
+
+    private final Map<String, String> offset;
+
+    public BinlogOffset(Map<String, String> offset) {
+        this.offset = offset;
+    }
+
+    public BinlogOffset(String filename, long position) {
+        this(filename, position, 0L, 0L, 0L, null, null);
+    }
+
+    public BinlogOffset(
+            String filename,
+            long position,
+            long restartSkipEvents,
+            long restartSkipRows,
+            long binlogEpochSecs,
+            @Nullable String restartGtidSet,
+            @Nullable Integer serverId) {
+        Map<String, String> offsetMap = new HashMap<>();
+        offsetMap.put(BINLOG_FILENAME_OFFSET_KEY, filename);
+        offsetMap.put(BINLOG_POSITION_OFFSET_KEY, String.valueOf(position));
+        offsetMap.put(EVENTS_TO_SKIP_OFFSET_KEY, String.valueOf(restartSkipEvents));
+        offsetMap.put(ROWS_TO_SKIP_OFFSET_KEY, String.valueOf(restartSkipRows));
+        offsetMap.put(TIMESTAMP_KEY, String.valueOf(binlogEpochSecs));
+        if (restartGtidSet != null) {
+            offsetMap.put(GTID_SET_KEY, restartGtidSet);
+        }
+        if (serverId != null) {
+            offsetMap.put(SERVER_ID_KEY, String.valueOf(serverId));
+        }
+        this.offset = offsetMap;
+    }
+
+    public Map<String, String> getOffset() {
+        return offset;
+    }
+
+    public String getFilename() {
+        return offset.get(BINLOG_FILENAME_OFFSET_KEY);
+    }
+
+    public long getPosition() {
+        return longOffsetValue(offset, BINLOG_POSITION_OFFSET_KEY);
+    }
+
+    public long getRestartSkipEvents() {
+        return longOffsetValue(offset, EVENTS_TO_SKIP_OFFSET_KEY);
+    }
+
+    public long getRestartSkipRows() {
+        return longOffsetValue(offset, ROWS_TO_SKIP_OFFSET_KEY);
+    }
+
+    public String getGtidSet() {
+        return offset.get(GTID_SET_KEY);
+    }
+
+    public long getTimestamp() {
+        return longOffsetValue(offset, TIMESTAMP_KEY);
+    }
+
+    public Long getServerId() {
+        return longOffsetValue(offset, SERVER_ID_KEY);
+    }
+
+    private long longOffsetValue(Map<String, ?> values, String key) {
+        Object obj = values.get(key);
+        if (obj == null) {
+            return 0L;
+        }
+        if (obj instanceof Number) {
+            return ((Number) obj).longValue();
+        }
+        try {
+            return Long.parseLong(obj.toString());
+        } catch (NumberFormatException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)